clang/lib/CConv/ConstraintBuilder.cpp

```diff
@@ -36,13 +36,26 @@ void processRecordDecl(RecordDecl *Declaration, ProgramInfo &Info,
         // We only want to re-write a record if it contains
         // any pointer types, to include array types.
         for (const auto &D : Definition->fields()) {
-          if (D->getType()->isPointerType() || D->getType()->isArrayType()) {
-            if(FL.isInSystemHeader() || Definition->isUnion()) {
-              CVarSet C = Info.getVariable(D, Context);
-              std::string Rsn = "External struct field or union encountered";
-              CB.constraintAllCVarsToWild(C, Rsn, nullptr);
+          bool mark_wild = ((D->getType()->isPointerType()
+                             || D->getType()->isArrayType()) &&
+                            (FL.isInSystemHeader() || Definition->isUnion()));
+          if (!mark_wild && inFunc) {
+            Decl *D = Declaration->getNextDeclInContext();
+            if (VarDecl *VD = dyn_cast<VarDecl>(D)) {
+              if (!VD->getType()->isPointerType()
+                  && !VD->getType()->isArrayType() && !VD->hasInit()) {
```

Comment:
This feels complicated. Is there a way to make the logic a little more evident?

If `mark_wild` is true then `D` is definitely a pointer or array type. Here you are interested in whether it's *not* `mark_wild` but then you don't know if that's because you are not dealing with a pointer/array type, or because you were but it was in a system header. I don't know how `Definition` relates to `D` ?

Maybe you can explain why you pulled out `mark_wild` as its own variable?

Reply:
I'm adding comments to it now and doing a little restructuring, so hopefully my latest commit (forthcoming) will explain everything.